internal/services/recommendation_poller.go

```diff
@@ -15,112 +16,126 @@ import (
 	"github.com/redhatinsights/ros-ocp-backend/internal/utils/kruize"
 )
 
-func FetchRecommendations(msg *kafka.Message, consumer_object *kafka.Consumer) {
+func commitKafkaMsg(msg *kafka.Message, consumer_object *kafka.Consumer) {
+	_, err := consumer_object.CommitMessage(msg)
+	if err != nil {
+		log.Error("unable to commit msg: ", err)
+	}
+}
+
+func requestAndSaveRecommendation(kafkaMsg types.RecommendationKafkaMsg, recommendationType string) bool {
+	log := logging.GetLogger()
+	experiment_name := kafkaMsg.Metadata.Experiment_name
+	maxEndTimeFromReport := kafkaMsg.Metadata.Max_endtime_report
+	var poll_cycle_complete bool = false
+
+	recommendation, err := kruize.Update_recommendations(experiment_name, maxEndTimeFromReport)
+	if err != nil {
+		end_interval := utils.ConvertDateToISO8601(maxEndTimeFromReport.String())
+		if err.Error() == fmt.Sprintf("Recommendation for timestamp - \" %s \" does not exist", end_interval) {
+			log.Infof("recommendation does not exist for timestamp - \" %s \"", end_interval)
+		}
+		log.Errorf("unable to list recommendation for: %v", err)
+	}
+
+	// TODO: Is_valid_recommendation to be called on every container record v20.1 upgrade on wards
+	if kruize.Is_valid_recommendation(recommendation) {
+		containers := recommendation[0].Kubernetes_objects[0].Containers
+		for _, container := range containers {
+			for _, v := range container.Recommendations.Data {
+				marshalData, err := json.Marshal(v)
+				if err != nil {
+					log.Errorf("unable to list recommendation for: %v", err)
+				}
+
+				// Create RecommendationSet entry into the table.
+				recommendationSet := model.RecommendationSet{
+					WorkloadID:          kafkaMsg.Metadata.Workload_id,
+					ContainerName:       container.Container_name,
+					MonitoringStartTime: v.Duration_based.Short_term.Monitoring_start_time,
+					MonitoringEndTime:   v.Duration_based.Short_term.Monitoring_end_time,
+					Recommendations:     marshalData,
+				}
+				if err := recommendationSet.CreateRecommendationSet(); err != nil {
+					log.Errorf("unable to save a record into recommendation set: %v. Error: %v", recommendationSet, err)
```

Comment:
I think if we are unable to save the recommendations then also we should not commit the message?

Reply:
I do agree, relocated poll_cycle_complete, feel free to take a look

Reply:
@saltgen if I see the whole code block now below is my understanding,
1. If CreateRecommendationSet() and CreateHistoricalRecommendationSet() are successful 
    - We commit the message.
2. If CreateRecommendationSet() and CreateHistoricalRecommendationSet() are unsuccessful 
    - We dont commit the message.
3. If CreateRecommendationSet() is successful and CreateHistoricalRecommendationSet() is unsuccessful
    - We don't commit the message
4. If CreateRecommendationSet() is unsuccessful and CreateHistoricalRecommendationSet() is successful
    - We don't commit the message

So in case 3 and 4 we are able to save the data in one of the table but not both. In this situation if we don't commit and when next time we would again handle same message,
A. If CreateRecommendationSet() failed in first iteration this means would be able to save it because the end time is still old in database.
B. If CreateHistoricalRecommendationSet() failed in first iteration this means the end timestamp in db is now same as kafka message and this mean we won't be able to do CreateHistoricalRecommendationSet().

Is it okay if we miss to save in HistoricalRecommendation table in this situation?

Reply:
If we wish to either success or fail for both then I think we can do https://gorm.io/docs/transactions.html

Reply:
DB transaction added